### ISSUE TITLE
Community(multiplayer) stats

### ISSUE THREAD

**Issue by syt0r** [SYtor/Kanji-Dojo#102]
There is a pretty fun feature of duolingo with weekly ranking based on XP users earned. I could do something similar, allowing user to send their stats anonymously (so that there won't be a need to manage accounts) and then display their results compared to community stats, making learning more engaging and competitive

It can be something simple like a text field with encouraging messaged "You're doing better than 60% of users", "You're in a top 99.9% users based on performance 😅" that would be based by some `score` = 40 * new + 10 * review, different for readings and writings, maybe some additional data like average count of kanji learnt/reviewed per day

**Comment by sl08154711** [comment]
A highscore? Yes! Absolutely! Just a bit of nitpicking: if the system was really anonymous then users could send in their stats multiple times. So you'd need something to identify them which is not anonymous. One should also make sure that users don't mess with the user database 😅 or have a look at the code and work out how to wreck things.
The score could be a rolling average over the last week to give new users a chance to "compete".
How about badges or achievements? Like "user has learned 10, 100 or 1000 characters" or "practiced continuously for 1, 5, 10 ... days".
Anyway, I wouldn't mind having to create an account but I also know how much work it is to manage user accounts and PII.

**Comment by syt0r** [comment]
Yeah, it can be easily tampered without any authentication. I don't know how anonymous it is but my idea was to generate random hash string to represent user. And in case someone try to manually feed api with fake data I could probably make a limit of entries per ip address or something. Honestly I wish I could simply use firebase for auth but then f-droid and maybe desktop versions wouldn't have this feature

**Comment by neithernut** [comment]
> Yeah, it can be easily tampered without any authentication. I don't know how anonymous it is but my idea was to generate random hash string to represent user.

With randomized user ids (such as UUIDs) it's generally best practice to rotate/regenerate them (client side) every once in a while if possible. IANAL but at least in some jurisdictions, you may also be legally required to let users opt in to such a stats-gathering feature or at least show them a disclaimer and maybe provide an opt-out.

> And in case someone try to manually feed api with fake data I could probably make a limit of entries per ip address or something.

Yes, you will definitely want to rate-limit such an API in some way (maybe in general _and_ per IP/user id). Depending on what you are using on the server-side, there probably already exists some off-the-shelf solution for that.

> Honestly I wish I could simply use firebase for auth but then f-droid and maybe desktop versions wouldn't have this feature

Imo that would also be reasonable?
